cmd/kaf/produce.go

```diff
@@ -87,6 +89,26 @@ func readFull(reader io.Reader, out chan []byte) {
 	close(out)
 }
 
+func valueEncoder() codec.Encoder {
+	if protoType != "" {
+		return proto.NewProtoCodec(protoType, reg)
+	} else if avroSchemaID != -1 {
+		return avro.NewAvroCodec(avroSchemaID, schemaCache)
+	} else {
+		return &codec.BypassCodec{}
+	}
+}
+
+func keyEncoder() codec.Encoder {
+	if keyProtoType != "" {
```

Comment:
this might be the time for a breaking change..maybe.

We could add:
--key-codec auto|proto|avro-json|avro 

auto would be default, and behave as-is. it would try its best to decide what do do: if proto key/value flags are given, it will try proto. if avro flags are given, it will try avro-json. but it can be set to specifically pick avro (see #195)

Reply:
I'm not sure about this one, for proto I think it would be a little redundant because you have to specify `--key-proto-type` anyways. I guess we don't have the same problem we have with avro here.

Maybe a flag specific for avro would be enough. It also doesn't need to be specific to keys, It can apply for both keys and values at the same time. Something like:
```
value | kaf produce --avro-key-schema-id x -- avro-schema-id y --key 'key' --avro-input [json|avro-json]
```

What do you think?

There's a separate issue with avro and proto though. The current behaviour is a little misleading imo. If you specify both proto and avro, proto takes precedence. Maybe we should have some validation and throw an error in this case.

Reply:
it's ok for me. do you know if "avro-json" is the right term? is there any "better" term in the avro world for that?

Reply:
Probably not, I'll check their docs see if I find anything. Another option would be just a boolean flag like `--avro-strict` then the docs can say something regarding being strict about unions.

A wiki page with examples of what we've been talking would be nice too.